Expression language for custom output fields

talbring opened the pull request:
## Proposed Changes
The is a first step in updating the output structure in order to make it a little bit more flexible and easier to maintain. The plan is to eventually have a consistent way of defining objective functions and/or define custom output quantities on individual surfaces.  

I introduced a new structure to store scalar quantities for the history output. This makes it much easier to filter and loop through a specific subset of fields. Next step is to have per surface values stored in the same structure. Furthermore, creating structure to compute problem-specific outputs like aerodynamic coefficients, time-averaged quantities, surface quantities and so on.

## Related Work

## PR Checklist
*Put an X by all that apply. You can fill this out after submitting the PR. If you have any questions, don't hesitate to ask! We want to help. These are a guide for you to know what the reviewers will be looking for in your contribution.*

- [x] I am submitting my contribution to the develop branch.
- [x] My contribution generates no new compiler warnings (try with the '-Wall -Wextra -Wno-unused-parameter -Wno-empty-body' compiler flags).
- [x] My contribution is commented and consistent with SU2 style.
- [ ] I have added a test case that demonstrates my contribution, if necessary.
- [ ] I have updated appropriate documentation (Tutorials, Docs Page, config_template.cpp) , if necessary.

talbring commented:
I am currently testing an expression parser to define custom fields in the config file. For this I modified the handling of the curly brackets {}. They now declare a "literal" interpretation, which means everything inside is considered as one string and will not be split while strings are tokenized when reading the config file. 

When you declare the output fields you can then use expressions involving all available fields, e.g.

`SCREEN_OUTPUT= INNER_ITER, RMS_DENSITY, {sqrt(LIFT)/DRAG**2}`

At the moment only inline expression are possible. But I am considering also to use a file to define more complex functions. The goal is to make that framework more flexibel to that we can use it also to define custom integrals over markers, functions defined on the whole flow field, custom boundary conditions, inlet conditions and so on. 

Important to note is that I already differentiated the parser, which means everything that is defined can be used also for the adjoint solver.

pcarruscag commented on SU2_CFD/src/output/CMultizoneOutput.cpp:
```diff
@@ -87,55 +87,39 @@ CMultizoneOutput::~CMultizoneOutput() {}
-        SetHistoryOutputValue(name, ZoneHistoryFields[ZoneHistoryNames[iField]].value);
+    const COutFieldCollection& ZoneHistoryFields = output[iZone]->GetHistoryFieldsAll();
 
+    for (auto field : ZoneHistoryFields.GetReferencesAll()){
```
I think auto type deduction looses "referenceness" i.e. you may want `auto& field` or `const auto& field` depending on the return type of GetReferencesAll

pcarruscag commented on Common/include/toolboxes/CIndexedMap.hpp:
```diff
@@ -0,0 +1,147 @@
+   * \param[in] key - The key that should be checked
+   * \return <TRUE> then when the key could be found, otherwise <FALSE>
+   */
+  bool FindKey(const Key& key) const {
```
Why not call it "CheckIsKey" or something like that? Find sounds like something more than a check

pcarruscag commented on Common/include/toolboxes/CIndexedMap.hpp:
```diff
@@ -0,0 +1,147 @@
+   */
+  const Item& GetItem(const int i) const {
+    return insertionVector[i]->second;
+  }
```
The code will not compile for `Key = int`, and it will have very surprising behavior if e.g. `Key = unsigned long`, you may want to disambiguate the names e.g. GetItemByIndex, GetNthItem

talbring commented:
@pcarruscag Thanks. I already incorporated your comments!

I extended the expressions to the volume output fields. Similar to the `SCREEN_OUTPUT` you can now define custom volume field variables directly in the `VOLUME_OUTPUT` option, e.g,

`VOLUME_OUTPUT= COORDINATES, SOLUTION, {MOMENTUM_X/DENSITY}`

This required to remove all `-` in the field names for obvious reasons.

Note that at the moment it is required that all fields used in the expression are also added manually to the output. I will probably make that automatic soon.

talbring commented:
It is now possible to define more complex functions in a separate file called `functions.su2x`. The functions defined there are available in the inline expressions. You can either give the functions explicit arguments or you can use them without any arguments. All fields are available in the global scope. Which variables are available depends on the context in which the function is used in the code. 

functions.su2x:
```
function dragOverLift(){
  l = LIFT;
  d = DRAG;
  return d/l;
}

function speedOfSound(gamma){
  p = PRESSURE;
  d = DENSITY;
  return sqrt(gamma*p/d);
}
```

config file:
```
SCREEN_OUTPUT=(INNER_ITER, WALL_TIME, RMS_DENSITY, {dragOverLift()})
VOLUME_OUTPUT = (COORDINATES, SOLUTION, {speedOfSound(1.4)})
```

The language is much more flexible and allows also for if-statements, loops, maps and so on. For the output it does not make that much sense, but for initial conditions it might be useful later.

jayantmukho commented:
> The language is much more flexible and allows also for if-statements, loops, maps and so on. For the output it does not make that much sense, but for initial conditions it might be useful later.

This could be useful even for general outputs. One example that comes to mind is if you wanted to calculate the momentum/pressure loss due to a shock. Excited for these new features!

jayantmukho commented on QuickStart/inv_NACA0012.cfg:
```diff
@@ -261,7 +261,7 @@ CONV_CAUCHY_ELEMS= 100
 CONV_CAUCHY_EPS= 1E-6
 %
-SCREEN_OUTPUT=(INNER_ITER, WALL_TIME, RMS_DENSITY, LIFT, DRAG, CAUCHY_SENS_PRESS, CAUCHY_DRAG RMS_ADJ_DENSITY RMS_ADJ_ENERGY)
+SCREEN_OUTPUT=(INNER_ITER, WALL_TIME, RMS_DENSITY, LIFT@airfoil, DRAG, CAUCHY_SENS_PRESS, CAUCHY_DRAG RMS_ADJ_DENSITY RMS_ADJ_ENERGY)
```
That's a cool feature. Would it be possible to specify multiple markers? Something along the lines of `LIFT@(upper_surface, lower_surface)

economon commented:
@talbring : great progress! As previously discussed, I think this goes in the correct direction.. I would love to see completely customizable obj/constraints/parameters in a flexible way, for any problem a user can dream up.

Like @clarkpede says, sometimes the objectives can require some heavy lifting (and require many local solution variables), and this thing is quite difficult to implement currently (it would be great if things like #889 were made easier). Do you think the language parser can eventually be used for this, or do you think we should just stick to writing those implementations in the COutput classes for each problem type (flow, elasticity, heat, etc).?

pcarruscag commented:
On this issue of custom outputs we should probably follow the industry, i.e. provide basic building blocks such as area/volume integrals, mass flow averages, etc. of arbitrary functions (functions of the output fields) and then on top of that structure build some useful defaults (e.g. force / moment coefficients).

talbring commented:
Here is a small description on how the user defined functions work at the moment : https://gist.github.com/talbring/cfd26e25218a15b8788ce4e047f97ed6. Please let me know what you think.

economon commented:
@talbring : since you are already working on the output here and if you agree, can you please add the sensitivities as a required output for adjoint runs even if the volume fields are customized (like we do for coords/conservatives for the primal)? I have gotten several reports that this is tripping up some folks, e.g., see #964

talbring commented:
> @talbring : since you are already working on the output here and if you agree, can you please add the sensitivities as a required output for adjoint runs even if the volume fields are customized (like we do for coords/conservatives for the primal)? I have gotten several reports that this is tripping up some folks, e.g., see #964 

In fact I already deactivated it in this branch. It makes no sense to allow customizable output for the adjoint anyway and it was giving troubles when you want to use the same config for primal and adjoint.

talbring commented:
Some update here: I had to get rid of the CParse library ... It was too slow since it is based  solely on RTTI. I replaced it with [exprtk](https://github.com/ArashPartow/exprtk). This is a fast and widely used expression library. The only drawback is that it is a header-only file with ~40k(!) lines of code. Compilation takes quite long, so I am using pimpl approach to interface it with SU2 in order not to recompile it for every compilation target. Furthermore I added the exprtk code as an archive (similar to what we do with the tecplot IO library).

pcarruscag commented:
Don't forget #1025 @talbring since you are moving some code into modules.

erangit commented:
Tim,
Stale is taking over this PR. I think it is useful, not only for the ventilator's effort. Would you like to pack at least the part that is completed and merge to develop?
Thxs,
Eran
